### ISSUE TITLE
Not getting propmted with 3D secure 2

### ISSUE THREAD

**Issue by adriamt** [Adyen/adyen-android#326]
Our app was working with adyen dropin already and now we have made some changes into our app.

If we made a transaction with a normal card, there is no problem and the charge is done fine. The problem occurs when we use a card with 3D secure 2.

After getting the payment methods from our server, passing them to the dropin, selecting a card (with 3D Secure 2) and introducing the CVC, we receive the `makePaymentsCall` call to our AdyenService and it's ok from our server, we get our response and it has `"resultCode":"IdentifyShopper"` but that's all, we are never prompted to challenge the 3D Secure 2.

Sometimes I made it work cleaning the project, or invalidating cache or something like that but after a few moments it's starts to fail again.

We were using implementation "com.adyen.checkout:drop-in:3.6.0 and we have tried 3.8.0.

Do you have any clue what it could be happening? It was working fine and now suddenly stopped working, our most relevant change since it was working it's maybe we have migrate to androidx

**Comment by caiofaustino** [comment]
So after getting `IdentifyShopper` you should also have an `action` object on that response that looks something like:
```JSON
"action": {
    "type": "threeDS2Fingerprint"
    "paymentMethodType": "scheme",
    "token":"<gibberish>",
    "paymentData":"<gibberish"
}
```
You need to get the that and pass it as as String as part of the  `CallResult`, something like this:
`CallResult(CallResult.ResultType.ACTION, "{type:threeDS2Fingerprint...}")`

**Comment by adriamt** [comment]
Yes, we have an action and we are passing it to the CallResult `CallResult(ResultType.ACTION, response.getJSONObject("action").toString())`

This is and example I've just done now

```
{
 "paymentData":"Ab02b4c0!BQABAgB8dtBcdFSS8e4Jo\/KJYUlX9eHk+L30P6Ds\/5qbSyhmxLT9fVxU4y37wgeCgziBq+dN+Mvj\/3tyw2KwtQlvoCy9hm2H\/Lo2CHOWY98DTyvTo8r8GTYCg7izXYec19ZPrQftMQ99wcFgApUaf3ZTEF40hwxY3nYE7BBYD0W1fsK6n6ee4hF0wJBLHJ6UooAZIl9QZOAl3ZtcJk0KPU6deYL+L2KxXze9j1wgjFNkLf4dkeTruim+sdMZYm9SzksEsF\/WBUe6Gwd4uomZOA0d0YTXAcRBPqmW0gdnYjanNzzkGVv7T4MXHLNVO4GEmK6J2aHe4IHHDmdj3vx9H85FMX0n4ylx7w5gPcSgArqGlOA9\/Jdw4BTtXuGBIeby2XzcKKCmDs\/vo\/cbOKdxOv\/k4fL6Ke3N4PKhEnLwlhhbuCsbBMn6GE\/iFwgCzokmQnAGhJ19B\/oXPOHTu3yx3do8OCnGYhqzBV7bH6\/DxZTCsnqGW+WDYJqYILXvo9sTCNoXDakZ2UD1\/erEuGZnLwiRO0ZO6T76LiuSqALmkbleaGhAGzlATpxZCyL0tJmz5CVxhNcdrt82Fnq571cZF4bT9IZVYGThaeg5VBaEjeac7a62kkwwoY1U8pocSSwsocxvJWkY4H197aSK+IHtAqN9HSS0Ii9qaz8\/lFTSKnVOK4fCYBCok+K1PtOvYEifl\/pVnpTdAEp7ImtleSI6IkFGMEFBQTEwM0NBNTM3RUFFRDg3QzI0REQ1MzkwOUI4MEE3OEE5MjNFMzgyM0Q2OERBQ0M5NEI5RkY4MzA1REMifaDVv+dUb+Er5\/Wt6dM\/MsuvMcWjvi1zVT\/rc0NlFUgjud0yMumC5bAZ7lWWXM72rDKjZUDYgJxEccZW9pRA\/+7hSC1ChJX47VKlXs0SaytAswZGhtnx6rwh7D4S1j+w6BucdTsOoc\/rm01mScw4fVffAKuTVYHE0iZbumL4UP\/HR0jUJYi7ZLzVCviqsYi7KagVZugOWcNUoo4NkPy2gKH1mY74aCW+N1jC3zD3bi72PoJicA\/xCacZATzi08loHu7zeYgep5JK9uQmO8f5z8e1AwaQNMSpSRl2kijHW7HmSVZQac6PsqsVYmePAY287oJiPuDffnCFEHJK5aOruGtAsJRPWgR2vpBFSP7E8a4mlZzuJ6m3m2Mp7rLXtyuQHzJ\/I2bpF4DMlE1gFgi3T75OlVhemLtmqk82Q5fY5Q+LGpBcCbTwNlpLFjNWxbHQ+fWhzZoC6SkN6xaLLnpENm1wnOwV9KLwX\/JevtoO\/dMGrsuKgYd3Zf0HRRfhykcXHepSkVTexFN2Bdiu13henFiQkN18H3ywGYzelQLxMnXuSbcpYuudH2r38YXoAUNfPy\/VSxn9fLFEmeQ2eeOe1akjtannf\/wwlM+RICtV81o1hI0PRlBS8ckCKFggobJ2HOpWs0hFyhRoHZvPJPOaJinHYasD\/cuzgDMKlI2vr0xUFjUANKIjm9E4PgLBE5T9+UBtlmXReSF0OM5EhWKI1w2OqkBYUKDU9Kty7ARndnzox4o8JkMkftsx8uwnCkRYlfFRk6kfyt99\/PbkfzCs5NXxaZvejlhd0a8H2rhtekbrQ7xkStlEl3HX7kUhfzRCxKyMKttwO2COh7hM2Iqjl7d8RWsaW5yj45WQOrc9ZZ3v4EEJCPYb6p1yBp3frDe4w0yYurzyC\/VLxa2tGoa03iYo5LH7Vwt\/xDD3dmdPkjJhhff4wLJH+TU60frBES1bAmRS+fILKuiadxDH+VXz8KV9gxw5cOoeG49S9pgP+uTaeY\/ZFcp30HA423CWmgF82K548oWYG8C2EoWpzZj9ZeDjNNylEULe4vpe6lif\/cI\/131Jsdge5MEVqKXEwppIiGbNjdwQZ40SOo+Q2oqv048GTEoP0FK\/AiVpbW367yfxqtnX\/euX0ig1e3g0PHqCQ3Mhn8o0Fh9UhJB7asjsCt\/qzw8B2vCC23thgEyO82YevdIYeipC8T5gT7sWUnbdh7QfGcNzTYh1iabJfpehNv0xl90h+s+cROD0wRUt\/qnCgv+WQKkw\/BAh+m42\/a33RGULEV3zpuXcsuoarLakm9PhJ\/Z3DIkpQI1gKW3gU\/PbhzC21ul9qvlh24aso+GxnGHJwSOb7biPYqs9DDBTulvLmu6vLtncFdaHtnADCXrDv4xfKv+k7wn36+6nDtNofC7VbppYt53lQM\/XGE1p\/KeRRM+K+yRzgsZnaxr0LK9H0FwreZQkX1atWxR7xZPEOXvHwTl7SFV3IcN2H7TwQqM14hY9KgjcERadpsJk0bq4Ne26JiqtuyAcbLaWZZHJbwwxhsxl3hV08azBSV\/0TxTa1qnVZE5icGMuvRX5X7C2DA3SGKIs5JK4M74CyJLZUmLtLw+lZU88oCTgQrwsi1DHDxdGN4I9HgpeRDqOG8RukhB9YwgRo4xPPTtO+fqd\/TsHzD3u8ZgTJ\/GKdSBMc58So0JbI33PUrKVPY4RNSfj\/qqtjcKv9WYmVzkfKJfK8QfKiF7stKJrzbx0l5sEAb48EGhixjzSQWQ\/DRNGF5w9XxW99PbmGDu8KRftfOK87cdxwEo6KQTmrTynW4jfrsWCm+OM+JBRHIjfGY4Sv2R\/1R+cu5hQ2JMtwjpFXBDtGUqiwJDIIZ7yEAKDAXonlTRqROU7JXUcJyj\/wBxQx3X6\/s6ZsyrkrFJtY934KJ+XDRckNL\/rs8EyxFbiKYPnjlkzanmht1HlN08l5h25fLoniQpm\/RQ7ecJ4Hy0PibDVzNm31sJSt1+P4oSeYO1hR4h0Mz0VG\/BVqB4CGXe9hNN6tLQEy1v8clvOZetw1KrybbnhQfmY5pfwW4XaeIdUui6RE9nEsXiPqD5jUTHRim9piUDHRCqiV2rS9DJ1B3egw02\/iXL9H6NK4MXXYE\/y2kbD3mUgUsLrjteSannWjeVNOr6TIqL+EOSN9xsdh0qtGohv712d2E5bDb3lYy9XqJz+ha04qBon\/lnmkiQBCEa8yvixyBAdXqVKg1BMigC7xpZXfDRA7QzvVsadaGN+7zgqMug3tWlHAjPOmpT+HZc5Ln7VCs00hF+B7ryCc5RgDtrdU7a0BL5TLAk8pOmTAEdGApapSO9V8WrF3XojuIVF+pHr4eU5\/UQinhDSadmumNlLetIi\/zctfiBzNUOAWT9J5rMwN2IXanju1EZGYtVhaWCsbeClCppwpytD243HsTR8eT84K9jYy1mM4z\/XIgd7vNFH0D0IBAUnX8GqQGSApTtkh8dH84ybn0AP1YZqNZRi7+RCB7AgVDWo0FW\/jr+nTSqo6kflUAb5pszBdeAUzHRc5tWanz10XWxgLcGbmHHz9cU17IUe6hiuSnhExZanKk4Z2HqP+L36krmm\/nuMBz9eeGMg+\/pXwBCzc71C9duawDMNE2RS3bZmyaD\/KdMBVo3SNfRJYhjjelM7rdxrpvKJ9e4NU2QZwyXkpd3GGuCqQStyJezU\/yPEs2ZeI8iP0r7TWDZUPdHjfT3f\/GwQlZ28x368sjJ99CGYXJRn99lZ7M2SnSc1hDvf4TPAp38GnlwYV44rMvGpvgnJ+8oKO2EaPEVwtdtQKrjs2IwQS5QkQfX+C8zwYBKuKIZO6da4SfeKfwvMeN1Bbnd6QCh1ofBiT9SP6Akp00e71UbLcwl5sIcaD3ClXkOjLEmoc\/hOEhYOnnhdrjLiMAV\/aQMB8ai\/AOWRAA1JcbbWvd8L+HYuAbv5Mp3FeUM6ocC5KCMLGdVW5IVmetjRvQsssJ8OFdYvhD2JDIR1QA7BQa7Sctxe5EpR\/iPiccE4GQ3Orla1J\/1hV5ccoHYQ9W0TDF+GT7jcGJNc3nc2BX\/cE8UqLbw5QmFC4\/Wq+0AwCV\/OOcihtqNCd8zAMgoDoE94h5BzGMHHAhxIcTLcLBBjxD9sahOHgxmeSbwzQaAhQecXQFNCtkNUyOop4VNJ4J0oHSAMnwRGAQqCV98fx+l\/cHtcQXdck1bjgSc8ADMbUFi1vLPD+Ie8eW\/7i86rB7fq5zN0CnvVQUdzoLiNj71e71M073P7TTh1uvBgkPH3u8eRRDp2iihFaaYz28p0QmFnI4\/OTq+J4X0ArMXVgGZ4H13Q8Y51yIvWcP+6QflQjK90zm\/fVHU7qwcJYtZV\/zSMSPwFHX97yqLEAwie9Zfg0x8wANXp+arWKYAQlvuA6Os490mueoTGWVe1lOJfelImOXySRO5fTtzoWuh\/Uwqi0GqlPdAifWuPriQ27dZgzJD6wP8irxgHiu2YIBhQT1yW3X4GSrfJjWeQaiZc6a4bVeHWsHkpfKkCnS71XdMA2OmKfT46MTWaMaewmYIY1bRIGK8wEPsvxf4m1rYtdoH6rtu9KbhTHJRXIdy9BB0QiOf67dnJmXoojTxQcq6wBNkAp2cvLEs6dNz86cK\/fu1zbgjVvc45KbWHWdcLozlz4DgugHueNhRPZUcwWOABI2ccgR7AqzNZWyIWjjZFLOw1CMX6HbIRjuvK7AejU8DdYlhzugwWFbj00u+jxs7fRUUONifhLTb5kSjFXpJ2PzqV44o114odb1R4rbn1hX88I0nlr6RBx6xh\/6xHBxllIzz6JnqFXAPIi\/bMnZpR22hRUGKkCjaBqmoyr+Jm+D9T5ZEutaEojwZS2OFb8Pjus+I+CvU6Wl6iavXGgVuhmMGkUxk819zkqPmXPztSrZpLupklLXrGRIMqFl5UEIuxmq7i1tfimWKi7rSaWWkGS7Kua5jiXHPSlq7AbETaVfI0Wb0\/SVLaIxRkw4PJAeRWAje0Bdc+r+A4VZ+XvqRMkvZVF3cPYqHcJcZlzqKt\/eDHYh\/dcqLAsQrKYBPWWln5fbOwq6OSVHDktcLfNy9Lfl+AdehRgVeqe0jd6w3qB98iSbg\/I\/+oZrmPC2ymD7bAVS9PXxOWGPnEfmhAVKLbNCrnFt\/xYNETys0Qv+PLEGLUU6JfEQVw5+W6oGku8rd+RGAnXLyGUWKFN1wyTrN5TZGCdmCHvmMOEWiWI3B3n4ZR7PItPia0PjSRdm0fksaPkpDYeKw47dsKLTbo7xxAGPgQ+Dlk65x8cyzP5vrP17VtqL0+SxLlEZ09KowwalnuetdwwGoZXy8yrd82\/YBSn0P5bGXRgOqVHuasJ0hsXNUAsou96wOmbQbZ2ewZKjmyk9tzld5p4CTBKmkC71DTnhs1dPFeizm982cetXtA2Y4uNT9UFqwm7QqJ8jQk\/m+J2knhfgacx8VTwtb7JkW8D1piEywWd6lOUOGeExCJGDUFvrOABNWlqTIIoTFMR9dkVKhV2C787otJCG7GNVRmkFpcI+S83BtCmkadyMU5MyZq2OgKDkIgGvtpp9185IYL67Q7hNYGY1Sa1Q\/7pAU+ADn7C+d7hAtqbBbgW19QYeiTd\/mh8ih2oahiTDWBwv5chdURMlZjcwC7AhCCyghzq2ElYizgcsJLjY\/Vnho\/xPrCoA\/\/syC6fJYL1G9PGgXPDIHTExhV\/3JrFEEKRNSL7VTLvL5DXGfc+z5+WUzn1+4tHM5aeJOcj8EFTus8PmxyrqzCR2g9io1haWfCn3NBE2ed8mzkn5XSKIXneT++QwYwghbCNA1RSaMPtlwQ0CR3EVk8EDZMoCEWe0RnaXG0sNTDHdvFhceBzLarhRiD8L+i7FbSK+Z1u0NXXF\/n3vyH9L6xF5dWFE+3QbSZqfREU9p7id1C+w==",
"paymentMethodType":"scheme",
"token":"eyJkaXJlY3RvcnlTZXJ2ZXJJZCI6IkYwMTMzNzEzMzciLCJkaXJlY3RvcnlTZXJ2ZXJQdWJsaWNLZXkiOiJleUpyZEhraU9pSlNVMEVpTENKbElqb2lRVkZCUWlJc0ltNGlPaUk0VkZCeFprRk9XazR4U1VFemNIRnVNa2RoVVZaaloxZzRMVXBXWjFZME0yZGlXVVJ0WW1kVFkwTjVTa1ZTTjNsUFdFSnFRbVF5YVRCRWNWRkJRV3BWVVZCWFZVeFpVMUZzUkZSS1ltOTFiVkIxYVhWb2VWTXhVSE4yTlRNNFVIQlJSbkV5U2tOYVNFUmthVjg1V1RoVlpHOWhibWxyVTA5NWMyTkhRV3RCVm1KSldIQTVjblZPU20xd1RUQndaMHM1Vkd4SlNXVkhZbEUzWkVKYVIwMU9RVkpMUVhSS2VUWTNkVmx2YlZwWFYwWkJiV3B3TTJkNFNEVnpOemRDUjJ4a2FFOVJVVmxRVEZkeWJEZHlTMHBMUWxVd05tMXRabGt0VUROcGF6azVNbXRQVVRORWFrMDJiSFIyV21OdkxUaEVUMlJDUjBSS1ltZFdSR0ZtYjI5TFVuVk5kMk5VVFhoRGRUUldZV3B5Tm1ReVprcHBWWGxxTlVZemNWQnJZbmc0V0RsNmExYzNVbWx4Vm5vMlNVMXFkRTU0TnpaaWNtZzNhVTlWZDJKaVdtb3hZV0Y2VkcxR1EyeEViMGR5WTJKeE9WODBObmNpZlE9PSIsInRocmVlRFNNZXNzYWdlVmVyc2lvbiI6IjIuMS4wIiwidGhyZWVEU1NlcnZlclRyYW5zSUQiOiJhN2E3MzE2MS02MWUxLTRlZGItOTUzYS04OGRmNmZlOWY2OTUifQ==",
"type":"threeDS2Fingerprint"
}

**Comment by adriamt** [comment]
I have some more info. With the version 3.6.6 I've just cleaned the project, invalidated cache and synced again.

I've compiled to my device again and now it's seems that work I get prompted with 3D Secure 2 (My device has Android 10)
I've compiled to a simulator again and it doesn't work (Simulator is an Android simulator with Android 9)
I've compiled to another devices and it doesn't work either (Device with Android 10)

**Comment by caiofaustino** [comment]
That looks really weird, can you add this line before starting the drop-in and check the logcat output?
I can't tell at what point the flow could be failing.
`Logger.setLogcatLevel(Log.DEBUG)`

**Comment by adriamt** [comment]
I've just did it and this is the result, after that I don't receive nothing.
```
D/CO.DropInService: handleCallResult - ACTION
D/CO.DropInActivity: callResultReceiver onReceive
    handleCallResult - ACTION
D/CO.Adyen3DS2Component: identifyShopper
D/CO.DropInService: onDestroy
D/CO.Adyen3DS2Component: initialize 3DS2 SDK

**Comment by caiofaustino** [comment]
Thanks a lot for your input. Doing some tests here I was able to find a bug of an exception that was being thrown in a background thread and there for was failing silently.

I'm not sure if it's the same one you found, but can you please try to initialize the `Adyen3DS2Component` with a configuration object with the protocol version specified? Like this
```kotlin
val config = Adyen3DS2Configuration.Builder(this@MainActivity)
    .setProtocolVersion("2.1.0")
    .build()
val threeds2Component = Adyen3DS2Component.PROVIDER.get(this@MainActivity, application, config)
```

**Comment by adriamt** [comment]
Where should I put thus configuration? I'm using dropin integration.
``` val cardConfiguration = CardConfiguration.Builder(activity, encryption_key).setShopperLocale(Locale.getDefault()).build()


            val amountObj = Amount()
            amountObj.currency = "EUR"
            amountObj.value = amount

            val dropInConfiguration = DropInConfiguration.Builder(activity, resultIntent, AdyenService::class.java)
                    // Optional. Use if you want to display the amount and currency on the Pay button.
                    .setAmount(amountObj)
                    // When you're ready to accept live payments, change the value to one of our live environments.
                    .setEnvironment(Environment.TEST)
                    // Optional. Use to set the language rendered in Drop-in, overriding the default device language setting. See list of Supported languages at https://github.com/Adyen/adyen-android/tree/master/card-ui-core/src/main/res
                    // Make sure that you have set the locale in the payment method configuration object as well.
                    .setShopperLocale(Locale.getDefault())
                    .addCardConfiguration(cardConfiguration)
                    .build()

            Logger.setLogcatLevel(Log.DEBUG)
            DropIn.startPayment(activity, paymentMethodsApiResponse, dropInConfiguration)

**Comment by caiofaustino** [comment]
Yeah sorry, I just noticed this and deleted the comment. I was trying to give you a work around for now, but I think you will have to wait for the fix. I'll work on it today and try to make a release by next week.

**Comment by caiofaustino** [comment]
Oh wait, are you using an alpha version of 4.0.0?

**Comment by adriamt** [comment]
No, I'm using 3.6.6

**Comment by adriamt** [comment]
Should I try 4.0.0 alpha?

**Comment by caiofaustino** [comment]
No, the issue I found only happens in the alpha version. Maybe it's something similar, but I'll investigate a bit more.

**Comment by adriamt** [comment]
Ok, thanks

**Comment by caiofaustino** [comment]
So like I mentioned in my deleted comment 😅, this seems to be an issue of an exception happening in a background thread and failing silently. I could make a fix with more logs or throwing the exception with a failure, but that wouldn't solve the problem, would just help finding out what it is.

So in order to kill two birds with one stone, could you maybe help me debug this so we can make a proper fix together with the extra logs?

What I'd need you to do is open the `Adyen3DS2Component` class that comes in the SDK, Android Studio should be able to open it from the dependency, and then put a break point in line 216.
https://github.com/Adyen/adyen-android/blob/3.6.6/3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.java#L216

After the debugger reaches that, you can enable breakpoints for uncaught exceptions. 
1. Click on View Breakpoints (Cmd+Shift+F8)
2. Enable Java Exception Breakpoints.
![Screen Shot 2021-01-21 at 13 54 35](https://user-images.githubusercontent.com/3094935/105354015-42725580-5bf0-11eb-8274-908964d46f7c.png)
3. Make sure you have Uncaught Exception enabled
4. Hit Resume on the Debugger
5. Check if any exception shows up, and share the stack trace with me please 😄 
Here is an example:
![Screen Shot 2021-01-21 at 13 57 09](https://user-images.githubusercontent.com/3094935/105354250-97ae6700-5bf0-11eb-9f9f-3bcc19caf5a4.png)

That would be incredibly hepful!

**Comment by adriamt** [comment]
This is what I get.
![image](https://user-images.githubusercontent.com/11394100/105365678-cb908900-5bfe-11eb-98de-471430016466.png)
I'm using the 3.6.6 actually and the exception occurs after the line 204 of Adyen3DS2Component.java
```
ThreeDS2Service.INSTANCE.initialize(context, configParameters, null, mUiCustomization);
````

After that I receive this exception, but I don't know if it's related.
![image](https://user-images.githubusercontent.com/11394100/105366019-288c3f00-5bff-11eb-8b5e-c27e59635dfb.png)

And thet I receive
![image](https://user-images.githubusercontent.com/11394100/105366207-5a9da100-5bff-11eb-9306-166fee2b0be5.png)
![image](https://user-images.githubusercontent.com/11394100/105366474-ad775880-5bff-11eb-8f09-c597906a4d72.png)

**Comment by caiofaustino** [comment]
Sorry but could you open up the stacktrace option inside the exception there on the right?

**Comment by adriamt** [comment]
Okei, here I go.
The first image is this stack trace
![image](https://user-images.githubusercontent.com/11394100/105459618-d3960a80-5c8a-11eb-97d6-934bb52859d1.png)
The second one
![image](https://user-images.githubusercontent.com/11394100/105459724-06d89980-5c8b-11eb-88eb-29dc732c74df.png)

And finally the third one
![image](https://user-images.githubusercontent.com/11394100/105459828-2e2f6680-5c8b-11eb-8b8e-07d1702990ed.png)

**Comment by caiofaustino** [comment]
This is super helpful, thanks a lot. The issue seems to be in the first one indeed. One last thing maybe can you please open the cause on the first one?

And also maybe can you share if you have `play-services-ads` or `play-services-location` as dependecies in your app?

**Comment by adriamt** [comment]
This is the cause of the first one
![image](https://user-images.githubusercontent.com/11394100/105460696-7733ea80-5c8c-11eb-883f-729100432ff6.png)

Yes, we have `play-services-location` but not the ads one

**Comment by caiofaustino** [comment]
Awesome, thanks a lot for the help! I will investigate with this info and get back to you if I find the root cause or a work around.

**Comment by adriamt** [comment]
Nice! Thanks a lot!! It would be perfect, because we are a bit blocked for this issue now hehehe. Thanks again.

**Comment by caiofaustino** [comment]
Thanks again, I think the real issue in the end is the 3rd exception. The 1rst one is just a caught exception we have to get the stacktrace, the second one seems to be internal to the Framework, but the third one I can see being the cause.
So it says the [`ErrnoException`](https://developer.android.com/reference/android/system/ErrnoException) happened when trying to read a file. That seems to be an error that happens in the native implementation of the device, so it's OEM dependant, which would explain why it was working in some devices and not others.

We definitely will make improvements on the logging and exception handling, but that might require a few updates that wouldn't be able to release so soon since you are blocked with this. The weird part to me is that is would be failing in most devices including the emulators, because that should've been caught by our tests.

In the mean time, are you making these tests in TEST environment? Maybe reach out to your contact here at Adyen and we can discuss an alternative, as in my experience currently most 3DS requests are still being done by 3DS1 redirect, so it could be a fallback to use for now while we investigate which devices this impacts and work on a fix. What do you think?

**Comment by adriamt** [comment]
Well with my device, sometimes it works and sometimes don't, so I think it's not related with the device itself.

Yes we are making this tests in the TEST environment, but we have a live one too.

**Comment by adriamt** [comment]
I've finally make it work with all devices I think. I've updated to version 4.0.0-alpha02 (I previously tested 4.0.0-alhpa01 and didn't work either)
And changed 
`CardConfiguration.Builder(activity, encryption_key).setShopperLocale(Locale.getDefault()).build()
`
To
`CardConfiguration.Builder(activity).setPublicKey(encryption_key).setClientKey(client_key).setEnvironment(Environment.TEST).setShopperLocale(Locale.getDefault()).build()`
And it works for now.

**Comment by caiofaustino** [comment]
Oh wow, I don't see how that seem to be related to the issue, but I'm glad it's working now. I was just finishing a fix that would at least fail the flow properly with feedback to the user instead of just spinning indefinitely. So if you want to test again after I release the fix we can see exactly what the problem was.

Thanks for the update though.

**Comment by adriamt** [comment]
Ok!! I'l try your fix when it's released to see the error. Thanks.

**Comment by caiofaustino** [comment]
Hey, we just released `3.8.1` if you want to take a look.

**Comment by jreij** [comment]
Hi @kightsonsanom thanks for reaching out! Which API version is your backend using? Also do you mind opening a new issue for this? I think it's similar but not the same 🙂

**Comment by kightsonsanom** [comment]
Thanks for quick reply @jreij :) It turns out we are using v52 API on backend, I believe this is the issue since it's quite stale. We are going to update to the latest version and if the issue is still there I will create a new ticket, thanks!

**Comment by jreij** [comment]
@kightsonsanom yes indeed, v52 with 4.5.0 might create lots of problems for you not just with 3DS2. We recommend v67+, let me know if the problem persists after upgrading!

**Comment by janw-ttc** [comment]
I am integrating Adyen payment for android with android drop-in using version 5.0.0 where i am able to start payment using DropIn and it is working fine but when it comes to 3d Secure authentication it doesn't do anything just showing loading . I tried to pass action object
 val action = Action.SERIALIZER.deserialize(jsonObj.getJSONObject("action"))
 DropInServiceResult.Action(action)
but it does not handle anything .

Below is the code snippet which i use to complete 3d secure authentication starting from creating DropinConfiguration object

 val cardConfiguration = CardConfiguration.Builder(this,Environment.TEST, BuildConfig.ADYEN_CLIENT_KEY)
                .setShowStorePaymentField(false)
                .setHolderNameRequired(true)

                .setHideCvc(false)
                .build()

            val dropInConfiguration = DropInConfiguration.Builder(this, Environment.TEST,BuildConfig.ADYEN_CLIENT_KEY)
                .setAmount(amount)
                .add3ds2ActionConfiguration(
                    Builder(Locale.US, Environment.TEST, BuildConfig.ADYEN_CLIENT_KEY).setThreeDSRequestorAppURL(RedirectComponent.getReturnUrl(this)).build(),
                )
                .addRedirectActionConfiguration(
                    RedirectConfiguration.Builder(Locale.US, Environment.TEST, BuildConfig.ADYEN_CLIENT_KEY).build(),
                )
                .addCardConfiguration(cardConfiguration)
                .build()


            val paymentMethodsApiResponse = paymentMethods?.let { it1 ->
                PaymentMethodsApiResponse.SERIALIZER.deserialize(
                    it1
                )
            }
            if (paymentMethodsApiResponse != null) {
                DropIn.startPayment(
                    this,
                    dropInLauncher,
                    paymentMethodsApiResponse,
                    dropInConfiguration,
                    AdyenDropInService::class.java,
                )
            }
 and when i get the response for payment and it contains action object i passed that to DropinServiceResult.Action like below
  val action = Action.SERIALIZER.deserialize(jsonObj.getJSONObject("action"))
  DropInServiceResult.Action(action)

It doesn't redirect me to anywhere where I can complete the 3d secure payment.

**Comment by janw-ttc** [comment]
@jreij 
I am integrating Adyen payment for android with android drop-in using version 5.0.0 where i am able to start payment using DropIn and it is working fine but when it comes to 3d Secure authentication it doesn't do anything just showing loading . I tried to pass action object
 val action = Action.SERIALIZER.deserialize(jsonObj.getJSONObject("action"))
 DropInServiceResult.Action(action)
but it does not handle anything .

Below is the code snippet which i use to complete 3d secure authentication starting from creating DropinConfiguration object

 val cardConfiguration = CardConfiguration.Builder(this,Environment.TEST, BuildConfig.ADYEN_CLIENT_KEY)
                .setShowStorePaymentField(false)
                .setHolderNameRequired(true)

                .setHideCvc(false)
                .build()

            val dropInConfiguration = DropInConfiguration.Builder(this, Environment.TEST,BuildConfig.ADYEN_CLIENT_KEY)
                .setAmount(amount)
                .add3ds2ActionConfiguration(
                    Builder(Locale.US, Environment.TEST, BuildConfig.ADYEN_CLIENT_KEY).setThreeDSRequestorAppURL(RedirectComponent.getReturnUrl(this)).build(),
                )
                .addRedirectActionConfiguration(
                    RedirectConfiguration.Builder(Locale.US, Environment.TEST, BuildConfig.ADYEN_CLIENT_KEY).build(),
                )
                .addCardConfiguration(cardConfiguration)
                .build()


            val paymentMethodsApiResponse = paymentMethods?.let { it1 ->
                PaymentMethodsApiResponse.SERIALIZER.deserialize(
                    it1
                )
            }
            if (paymentMethodsApiResponse != null) {
                DropIn.startPayment(
                    this,
                    dropInLauncher,
                    paymentMethodsApiResponse,
                    dropInConfiguration,
                    AdyenDropInService::class.java,
                )
            }
 and when i get the response for payment and it contains action object i passed that to DropinServiceResult.Action like below
  val action = Action.SERIALIZER.deserialize(jsonObj.getJSONObject("action"))
  DropInServiceResult.Action(action)

It doesn't redirect me to anywhere where I can complete the 3d secure payment.

**Comment by jreij** [comment]
@janw-ttc thanks for reaching out. Please open a new issue for this to avoid notifying everyone on this thread, we'll be happy to help you there 🙂
